### PR TITLE
SONAR-8790 Add last analysis date on component related WS

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java [hunk 1]
```diff
@@ -195,25 +200,22 @@ private void checkPermissions(ComponentDto baseComponent) {
...
-
+    response.setBaseComponent(toWsComponent(baseComponent, organizationDto, referenceComponentsByUuid, lastAnalysis));
+    components.forEach(component -> response.addComponents(toWsComponent(component, organizationDto, referenceComponentsByUuid, Optional.empty())));
     return response.build();
   }
 
   private static WsComponents.Component.Builder toWsComponent(ComponentDto component, OrganizationDto organizationDto,
-    Map<String, ComponentDto> referenceComponentsByUuid) {
-    WsComponents.Component.Builder wsComponent = componentDtoToWsComponent(component, organizationDto);
+    Map<String, ComponentDto> referenceComponentsByUuid, Optional<SnapshotDto> lastAnalysis) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java]: MINOR Specify a "SnapshotDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java [hunk 2]
```diff
@@ -195,25 +200,22 @@ private void checkPermissions(ComponentDto baseComponent) {
   }
 
   private static TreeWsResponse buildResponse(ComponentDto baseComponent, OrganizationDto organizationDto, List<ComponentDto> components,
-    Map<String, ComponentDto> referenceComponentsByUuid, Paging paging) {
+    Map<String, ComponentDto> referenceComponentsByUuid, Optional<SnapshotDto> lastAnalysis, Paging paging) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java]: MINOR Specify a "SnapshotDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ShowAction.java [hunk 3]
```diff
@@ -122,4 +108,21 @@ private ComponentDto getComponentByUuidOrKey(DbSession dbSession, ShowWsRequest
     userSession.checkComponentPermission(UserRole.USER, component);
     return component;
   }
+
+  private static ShowWsResponse buildResponse(ComponentDto component, OrganizationDto organizationDto, List<ComponentDto> orderedAncestors, Optional<SnapshotDto> lastAnalysis) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ShowAction.java]: MINOR Specify a "SnapshotDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java [hunk 4]
```diff
@@ -20,28 +20,31 @@
...
+import static org.sonar.api.utils.DateUtils.formatDateTime;
 import static org.sonar.core.util.Protobuf.setNullable;
 
 class ComponentDtoToWsComponent {
   private ComponentDtoToWsComponent() {
     // prevent instantiation
   }
 
-  static WsComponents.Component.Builder componentDtoToWsComponent(ComponentDto dto, OrganizationDto organizationDto) {
+  static WsComponents.Component.Builder componentDtoToWsComponent(ComponentDto dto, OrganizationDto organizationDto, Optional<SnapshotDto> lastAnalysis) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: MINOR Specify a "SnapshotDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java [hunk 5]
```diff
@@ -20,28 +20,31 @@
...
     checkArgument(
       Objects.equals(dto.getOrganizationUuid(), organizationDto.getUuid()),
       "OrganizationUuid (%s) of ComponentDto to convert to Ws Component is not the same as the one (%s) of the specified OrganizationDto",
       dto.getOrganizationUuid(), organizationDto.getUuid());
-    return componentDtoToWsComponent(dto, organizationDto.getKey());
+    return componentDtoToWsComponent(dto, organizationDto.getKey(), lastAnalysis);
   }
 
-  private static WsComponents.Component.Builder componentDtoToWsComponent(ComponentDto dto, String organizationDtoKey) {
+  private static WsComponents.Component.Builder componentDtoToWsComponent(ComponentDto dto, String organizationDtoKey, Optional<SnapshotDto> lastAnalysis) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: MINOR Specify a "SnapshotDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: this rule should be challenged or a better alternative should be suggested

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java [hunk 6]
```diff
@@ -51,6 +54,7 @@ private ComponentDtoToWsComponent() {
     setNullable(emptyToNull(dto.path()), wsComponent::setPath);
     setNullable(emptyToNull(dto.description()), wsComponent::setDescription);
     setNullable(emptyToNull(dto.language()), wsComponent::setLanguage);
+    lastAnalysis.ifPresent(analysis -> wsComponent.setAnalysisDate(formatDateTime(analysis.getCreatedAt())));
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: 👍

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Pull request merged**